rust/lsp-lib/src/lsp_plugin.rs

```diff
@@ -0,0 +1,271 @@
+
+    fn update(
+        &mut self,
+        view: &mut View<Self::Cache>,
+        delta: Option<&RopeDelta>,
+        _edit_type: String,
+        _author: String,
+    ) {
+        if let Some(d) = delta {
+
+            let mut ls_client = self.0.lock().unwrap();
+            let sync_kind = ls_client.get_sync_kind();
+
+            let changes = match sync_kind {
+                TextDocumentSyncKind::None => return,
+                TextDocumentSyncKind::Full => {
+                    let text_document_content_change_event = TextDocumentContentChangeEvent {
+                        range: None,
+                        range_length: None,
+                        text: view.get_document().unwrap(),
+                    };
+                    vec![text_document_content_change_event]
+                }
+                TextDocumentSyncKind::Incremental => match get_document_content_changes(d, view) {
+                    Ok(result) => result,
+                    Err(err) => {
+                        eprintln!("Error: {:?} Occured. Sending Whole Doc", err);
+                        let text_document_content_change_event = TextDocumentContentChangeEvent {
+                            range: None,
+                            range_length: None,
+                            text: view.get_document().unwrap(),
+                        };
+                        vec![text_document_content_change_event]
+                    }
+                },
+            };
+
+            ls_client.send_did_change(view.get_id(), changes, view.rev);
+        }
+    }
+
+    fn did_save(&mut self, view: &mut View<Self::Cache>, _old: Option<&Path>) {
+        eprintln!("saved view {}", view.get_id());
+
+        let document_text = view.get_document().unwrap();
+        let mut ls_client = self.0.lock().unwrap();
+        ls_client.send_did_save(view.get_id(), document_text);
+    }
+
+    fn did_close(&mut self, view: &View<Self::Cache>) {
+        eprintln!("close view {}", view.get_id());
+    }
+
+    fn new_view(&mut self, view: &mut View<Self::Cache>) {
+        eprintln!("new view {}", view.get_id());
+
+        let document_text = view.get_document().unwrap();
+        let path = view.get_path().clone();
+
+        let view_id = view.get_id().clone();
+
+        if let Some(file_path) = path {
+            let extension = file_path.extension().unwrap().to_str().unwrap().to_string();
+
+            let mut ls_client = self.0.lock().unwrap();
+
+            if ls_client.file_extensions.contains(&extension) {
+                let document_uri =
+                    Url::parse(format!("file://{}", file_path.to_str().unwrap()).as_ref()).unwrap();
+
+                let workspace_root = ls_client.get_workspace_root(file_path);
+
+                let workspace_uri = match workspace_root {
+                    Some(path) => {
+                        Some(Url::parse(format!("file://{}", path.to_str().unwrap()).as_ref()).unwrap())
```

Comment:
This (in general) does not fail I guess, but it would make sense to return workspace uri from the method entirely and handle this failure in that method. Would do.